egui_glium/src/painter.rs

```diff
@@ -1,192 +1,386 @@
+    /// This is the format glow likes.
+    pixels: Vec<u8>,
+    pixels_res: (usize, usize),
 
-    /// Lazily uploaded from [`Self::pixels`],
-    /// or owned by the user via `register_native_texture`.
-    gl_texture: Option<Rc<SrgbTexture2d>>,
+    /// Lazily uploaded
+    gl_texture: Option<glow::NativeTexture>,
 }
 
-impl Painter {
-    pub fn new(facade: &dyn glium::backend::Facade) -> Painter {
-        let program = program! {
-            facade,
-            120 => {
-                vertex: include_str!("shader/vertex_120.glsl"),
-                fragment: include_str!("shader/fragment_120.glsl"),
-            },
-            140 => {
-                vertex: include_str!("shader/vertex_140.glsl"),
-                fragment: include_str!("shader/fragment_140.glsl"),
-            },
-            100 es => {
-                vertex: include_str!("shader/vertex_100es.glsl"),
-                fragment: include_str!("shader/fragment_100es.glsl"),
-            },
-            300 es => {
-                vertex: include_str!("shader/vertex_300es.glsl"),
-                fragment: include_str!("shader/fragment_300es.glsl"),
-            },
+#[derive(Copy, Clone, Debug)]
+#[allow(dead_code)]
+pub enum ShaderVersion {
+    Gl120,
+    Gl140,
+    Es100,
+    Es300,
+}
+
+impl ShaderVersion {
+    fn get(gl: &glow::Context) -> Self {
+        let glsl_ver = unsafe { gl.get_parameter_string(glow::SHADING_LANGUAGE_VERSION) };
+        let start = glsl_ver.find(|c| char::is_ascii_digit(&c)).unwrap();
+        let es = glsl_ver[..start].contains("ES");
+        let ver = glsl_ver[start..].splitn(2, ' ').next().unwrap();
+        let [maj, min]: [u8; 2] = ver
+            .splitn(3, '.')
+            .take(2)
+            .map(|x| x.parse().unwrap_or_default())
+            .collect::<Vec<u8>>()
+            .try_into()
+            .unwrap();
+        if es {
+            if maj >= 3 {
+                Self::Es300
+            } else {
+                Self::Es100
+            }
+        } else if maj > 1 || (maj == 1 && min >= 40) {
+            Self::Gl140
+        } else {
+            Self::Gl120
         }
-        .expect("Failed to compile shader");
+    }
 
-        Painter {
-            program,
-            egui_texture: None,
-            egui_texture_version: None,
-            user_textures: Default::default(),
+    fn version(&self) -> &'static str {
+        match self {
+            Self::Gl120 => "#version 120\n",
+            Self::Gl140 => "#version 140\n",
+            Self::Es100 => "#version 100\n",
+            Self::Es300 => "#version 300 es\n",
         }
     }
+}
+
+impl Painter {
+    pub fn new(gl: &glow::Context) -> Painter {
+        let header = ShaderVersion::get(gl).version();
+        let v_src = header.to_owned() + VERT_SRC;
+        let f_src = header.to_owned() + FRAG_SRC;
+        // TODO error handling
+        unsafe {
+            let v = gl.create_shader(glow::VERTEX_SHADER).unwrap();
```

Comment:
About errors: for now I think it is fine to panic on errors, but not to ignore errors.